runelite-client/src/main/java/net/runelite/client/plugins/puzzlesolver/solver/pathfinding/IDAStarMM.java

```diff
@@ -0,0 +1,838 @@
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.puzzlesolver.solver.pathfinding;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import net.runelite.api.Point;
+import static net.runelite.client.plugins.puzzlesolver.solver.PuzzleSolver.BLANK_TILE_VALUE;
+import static net.runelite.client.plugins.puzzlesolver.solver.PuzzleSolver.DIMENSION;
+import net.runelite.client.plugins.puzzlesolver.solver.PuzzleState;
+import net.runelite.client.plugins.puzzlesolver.solver.heuristics.Heuristic;
+
+public class IDAStarMM extends IDAStar
+{
+	private PuzzleState currentState;
+	private List<PuzzleState> stateList = new ArrayList<>();
+	private List<List<Integer>> validRowNumbers = new ArrayList<>();
+	private List<List<Integer>> validColumnNumbers = new ArrayList<>();
+
+	public IDAStarMM(Heuristic heuristic)
+	{
+		super(heuristic);
+
+		//Add valid numbers for rows and columns
+		validRowNumbers.add(Arrays.asList(0, 1, 2, 3, 4));
+		validColumnNumbers.add(Arrays.asList(5, 10, 15, 20));
+	}
+
+	@Override
+	public List<PuzzleState> computePath(PuzzleState root)
+	{
+		currentState = new PuzzleState(root);
+		stateList.add(new PuzzleState(root));
+
+		List<PuzzleState> path = new ArrayList<>();
+
+		//Reduce to 4x5
+		solveRow();
+
+		//Reduce to 4x4
+		solveColumn();
+
+		//Remove last state
+		stateList.remove(stateList.size() - 1);
+
+		//Pathfinder for 4x4
+		path.addAll(super.computePath(new PuzzleState(currentState)));
+
+		path.addAll(0, stateList);
+
+		return path;
+	}
+
+	private void solveRow()
+	{
+		int row = 0;
+		validateInput(row);
+
+		for (int i = row; i < DIMENSION; i++)
+		{
+			int valTarget = row * DIMENSION + i;
+
+			int valCurrent = currentState.getPiece(i, row);
+
+			if (valCurrent != valTarget)
+			{
+				Point locTarget = new Point(i, row);
+				moveTowardsVal(valTarget, locTarget,  true);
+			}
+		}
+	}
+
+	private void solveColumn()
+	{
+		int column = 0;
+		validateInput(column);
+
+		for (int i = column + 1; i < DIMENSION; i++)
+		{
+			int valTarget = column + i * DIMENSION;
+
+			int valCurrent = currentState.getPiece(column, i);
+
+			if (valCurrent != valTarget)
+			{
+				Point locTarget = new Point(column, i);
+				moveTowardsVal(valTarget, locTarget,  false);
+			}
+		}
+	}
+
+	private void validateInput(int val)
```

Comment:
This method is only ever used when ``val = 0``, it doesn't seem to be needed?

Reply:
It was just there to ensure that when solveRow or solveColumn was parameterized, then number entered as parameter was valid.